### DISCUSSION TITLE
Means of mocking appropriate context/methods for testing subclasses of `XunitTestCase`?

### DISCUSSION THREAD

**Discussion by Kritner** [xunit/xunit#2656] [Question] [answered]
Is there an easy way to do this?  I'm trying to (slowly) mock out (with `Moq`) the entire chain of calls that are needed when trying to write tests against subclasses of `XunitTestCase` and eventually logic within `RunAsync` itself.

Any thoughts/examples of this?

I'm "further than I was" at the beginning, but I feel like someone's had to have done this before and/or there are already existing helpers out there:

```cs
private readonly Mock<IXunitSerializationInfo> _xunitSerializationInfo = new();
private readonly Mock<ITestMethod> _testMethod = new();

    public MyTestClass()
    {
        _xunitSerializationInfo
            .Setup(s => s.GetValue<string>("DefaultMethodDisplay"))
            .Returns("Method");
        _xunitSerializationInfo
            .Setup(s => s.GetValue<string>("DefaultMethodDisplayOptions"))
            .Returns("None");
        _xunitSerializationInfo
            .Setup(s => s.GetValue<ITestMethod>("TestMethod"))
            .Returns(_testMethod.Object);
        
        var methodInfo = new Mock<IMethodInfo>();
        methodInfo.Setup(s => s.GetCustomAttributes(typeof(FactAttribute).AssemblyQualifiedName))
            .Returns(new List<IAttributeInfo>()
            {
                new Mock<IAttributeInfo>().Object
            });
        methodInfo.Setup(s => s.Name)
            .Returns("whatever");
        
        _testMethod
            .SetupGet(s => s.Method)
            .Returns(methodInfo.Object);
        var testClass = new Mock<ITestClass>();
        testClass
            .SetupGet(s => s.Class)
            .Returns(new Mock<ITypeInfo>().Object);
        _testMethod
            .SetupGet(s => s.TestClass)
            .Returns(testClass.Object);   
    }
```

**Comment by bradwilson** [comment, marked as answer]
We use NSubstitute rather than Moq, but you could use our mocks as a starting point. We obviously have a lot of testing around the execution pipeline.

https://github.com/xunit/xunit/blob/v2/test/test.utility/TestDoubles/Mocks.cs

**Reply by Kritner** [reply]
Oh nice, I'll give it a go, thanks!

**Reply by Kritner** [reply]
Are these substitutes exposed via a package already?  I've tried including a bunch of xunit.* packages in my solution to try and find the `Mocks` static class to no avail 😅.  

I've always found it a little tricky to determine which "package" a specific set of functionality/namespaces was exposed from when multiple packages are being built from a single repo. 

Or is this set of substitutes *not* exposed via package, and I should just copy them for my use?

**Reply by bradwilson** [reply]
> Are these substitutes exposed via a package already?

No, they were written for our unit tests only. While they may be useful as-is, they are not tested or supported outside of our particular usage.

**Reply by Kritner** [reply]
> No, they were written for our unit tests only. While they may be useful as-is, they are not tested or supported outside of our particular usage.

understood, I got my tests going with the help of the `Mocks` you pointed out.

Thanks!
